targets/TARGET_NORDIC/TARGET_NRF5/trng_api.c

```diff
@@ -37,54 +37,139 @@
  */
 
 #if defined(DEVICE_TRNG)
-#include "trng_api.h"
+
+#include "hal/trng_api.h"
+#include "hal/lp_ticker_api.h"
+
 #include "nrf_drv_rng.h"
 
+#define DEFAULT_TIMEOUT_US (1000*1000)
+
+/* Macro for testing if the SoftDevice is active, regardless of whether the 
+ * application is build with the SoftDevice or not.
+ */
+#if defined(SOFTDEVICE_PRESENT)
+#include "nrf_sdm.h"
+static uint8_t wrapper(void) {
+    uint8_t softdevice_is_enabled;
+    ret_code_t result = sd_softdevice_is_enabled(&softdevice_is_enabled);
+    return ((result == NRF_SUCCESS) && (softdevice_is_enabled == 1));
+}
+#define NRF_HAL_SD_IS_ENABLED() wrapper()
+#else
+#define NRF_HAL_SD_IS_ENABLED() 0
+#endif
+
+/** Initialize the TRNG peripheral
+ *
+ * @param obj The TRNG object
+ */
 void trng_init(trng_t *obj)
 {
     (void) obj;
 
-    (void)nrf_drv_rng_init(NULL);
+    /* Initialize low power ticker. Used for timeouts. */
+    static bool first_init = true;
+
+    if (first_init) {
+        first_init = false;
+        lp_ticker_init();
+    }
+
+    /* Initialize SDK RNG once. */
+    static bool trng_init = true;
+
+    if (trng_init && NRF_HAL_SD_IS_ENABLED()) {
+        trng_init = false;
+        nrf_drv_rng_init(NULL);
+    }
 }
 
+/** Deinitialize the TRNG peripheral
+ *
+ * @param obj The TRNG object
+ */
 void trng_free(trng_t *obj)
 {
     (void) obj;
-
-    nrf_drv_rng_uninit();
```

Comment:
Why has this been removed even in case SD is enabled?

Reply:
There is no good way (at least in this version of the SDK) to keep track of SoftDevice state. Imagine a case where `trng_init` is called before the SoftDevice is enabled and `trng_free` is called after, then we would have an inconsistent pairing of init/uninit calls.

In this version the TRNG is kept running after the first call. The power consumption for that is orders of magnitude lower than the SoftDevice in sleep mode.

Reply:
@marcuschangarm Ok, but with the present code, issuing multiple `trng_init / trng_free` calls leads to failure, which is not consistent with the TRNG API. What about calling `nrf_drg_rng_uninit` if only if `trng_init == false` -- i.e. if the TRNG has been initialized when SoftDevice was enabled -- and resetting it to `trng_init = true`? This way, multiple `trng_init / trng_free` calls would be fine, as long as it doesn't happen that the SoftDevice gets disabled in the middle of a `trng_init / trng_free` pair.

Reply:
> issuing multiple trng_init / trng_free calls leads to failure

I don't understand, can you please elaborate? `trng_init` doesn't do anything after the first successful call and `trng_free` is empty.

Reply:
@marcuschangarm There was a misunderstanding on my side, I agree that multiple `trng_init` calls with SoftDevice enabled are fine, with all but the first one being no-ops. 

Regarding the case you described -- SoftDevice being enabled after `trng_init` was called: Can we use `trng_init == false` (and perhaps give it a more descriptive name, pointing out that it's about TRNG initialization in the presence of SoftDevice) instead of `NRF_HAL_SD_IS_ENABLED()` in `trng_get_bytes` to avoid using SoftDevice functionality without being sure that it is properly initialized?

Reply:
> Can we use trng_init == false (and perhaps give it a more descriptive name, pointing out that it's about TRNG initialization in the presence of SoftDevice) instead of NRF_HAL_SD_IS_ENABLED() in trng_get_bytes to avoid using SoftDevice functionality without being sure that it is properly initialized?

The SoftDevice does its own TRNG initialization. So we might end up in a situation where we use the low-level TRNG API and steal random numbers from the SoftDevice.

How about moving `nrf_drv_rng_init` to `trng_get_bytes` instead?

```
int trng_get_bytes(trng_t *obj, uint8_t *output, size_t length, size_t *output_length)
{
    (void) obj;

    /* Use SDK RNG driver if SoftDevice is enabled. */
    if (NRF_HAL_SD_IS_ENABLED() ) {
        static bool nordic_driver_init = true;

        if (nordic_driver_init) {
            nordic_driver_init = false;
            nrf_drv_rng_init(NULL);
        }
```